### ISSUE TITLE
Broken opencv2 cmake config in hydro

### ISSUE THREAD

**Issue by meyerj** [ros-perception/vision_opencv#39]
With the recent update of the [opencv2](http://wiki.ros.org/opencv2) package from 2.4.6 to 2.4.9 in https://github.com/ros/rosdistro/commit/0ee098e3c707470ceabc4e9a624037d189204c0e image_geometry (and probably other packages) do not export the opencv2 libraries correctly in its catkin-generated config files. That's why several packages that depend on opencv2 indirectly via another package currently have linker errors in Jenkins, for example:
- http://jenkins.ros.org/job/ros-hydro-hector-object-tracker_binarydeb_precise_i386/49/
- http://jenkins.ros.org/view/HbinP32/job/ros-hydro-opt-camera_binarydeb_precise_i386/39/
- http://jenkins.ros.org/view/HbinP32/job/ros-hydro-pano-core_binarydeb_precise_i386/54/
- ...

The cmake config of OpenCV is different for 2.4.6 and 2.4.9 and this results in a different pkg-config/cmake config of package image_geometry:
- With `ros-hydro-opencv2-2.4.6-3precise-20140303-2244-+0000` and `ros-hydro-image-geometry-1.10.17-1precise-20140521-1853-+0000` from the public repo:
  
  ```
  $ cmake --find-package -DNAME=OpenCV -DCOMPILER_ID=GNU -DLANGUAGE=CXX -DMODE=LINK
  -rdynamic -ldebug /opt/ros/hydro/lib/libopencv_calib3d.so -ldebug /opt/ros/hydro/lib/libopencv_contrib.so -ldebug /opt/ros/hydro/lib/libopencv_core.so -ldebug /opt/ros/hydro/lib/libopencv_features2d.so -ldebug /opt/ros/hydro/lib/libopencv_flann.so -ldebug /opt/ros/hydro/lib/libopencv_gpu.so -ldebug /opt/ros/hydro/lib/libopencv_highgui.so -ldebug /opt/ros/hydro/lib/libopencv_imgproc.so -ldebug /opt/ros/hydro/lib/libopencv_legacy.so -ldebug /opt/ros/hydro/lib/libopencv_ml.so -ldebug /opt/ros/hydro/lib/libopencv_nonfree.so -ldebug /opt/ros/hydro/lib/libopencv_objdetect.so -ldebug /opt/ros/hydro/lib/libopencv_photo.so -ldebug /opt/ros/hydro/lib/libopencv_stitching.so -ldebug /opt/ros/hydro/lib/libopencv_superres.so -ldebug /opt/ros/hydro/lib/libopencv_ts.so -ldebug /opt/ros/hydro/lib/libopencv_video.so -ldebug /opt/ros/hydro/lib/libopencv_videostab.so -loptimized /opt/ros/hydro/lib/libopencv_calib3d.so -loptimized /opt/ros/hydro/lib/libopencv_contrib.so -loptimized /opt/ros/hydro/lib/libopencv_core.so -loptimized /opt/ros/hydro/lib/libopencv_features2d.so -loptimized /opt/ros/hydro/lib/libopencv_flann.so -loptimized /opt/ros/hydro/lib/libopencv_gpu.so -loptimized /opt/ros/hydro/lib/libopencv_highgui.so -loptimized /opt/ros/hydro/lib/libopencv_imgproc.so -loptimized /opt/ros/hydro/lib/libopencv_legacy.so -loptimized /opt/ros/hydro/lib/libopencv_ml.so -loptimized /opt/ros/hydro/lib/libopencv_nonfree.so -loptimized /opt/ros/hydro/lib/libopencv_objdetect.so -loptimized /opt/ros/hydro/lib/libopencv_photo.so -loptimized /opt/ros/hydro/lib/libopencv_stitching.so -loptimized /opt/ros/hydro/lib/libopencv_superres.so -loptimized /opt/ros/hydro/lib/libopencv_ts.so -loptimized /opt/ros/hydro/lib/libopencv_video.so -loptimized /opt/ros/hydro/lib/libopencv_videostab.so -ldebug -Wl,-rpath,/opt/ros/hydro/lib 
  ```
  
  ```
  $ cat /opt/ros/hydro/lib/pkgconfig/image_geometry.pc
  prefix=/opt/ros/hydro
  
  Name: image_geometry
  Description: Description of image_geometry
  Version: 1.10.17
  Cflags: -I/opt/ros/hydro/include -I/opt/ros/hydro/include/opencv
  Libs: -L/opt/ros/hydro/lib -limage_geometry -l:/opt/ros/hydro/lib/libopencv_calib3d.so -l:/opt/ros/hydro/lib/libopencv_contrib.so -l:/opt/ros/hydro/lib/libopencv_core.so -l:/opt/ros/hydro/lib/libopencv_features2d.so -l:/opt/ros/hydro/lib/libopencv_flann.so -l:/opt/ros/hydro/lib/libopencv_gpu.so -l:/opt/ros/hydro/lib/libopencv_highgui.so -l:/opt/ros/hydro/lib/libopencv_imgproc.so -l:/opt/ros/hydro/lib/libopencv_legacy.so -l:/opt/ros/hydro/lib/libopencv_ml.so -l:/opt/ros/hydro/lib/libopencv_nonfree.so -l:/opt/ros/hydro/lib/libopencv_objdetect.so -l:/opt/ros/hydro/lib/libopencv_photo.so -l:/opt/ros/hydro/lib/libopencv_stitching.so -l:/opt/ros/hydro/lib/libopencv_superres.so -l:/opt/ros/hydro/lib/libopencv_video.so -l:/opt/ros/hydro/lib/libopencv_videostab.so
  Requires: sensor_msgs
  ```
- With `ros-hydro-opencv2-2.4.9-1precise-20140613-1754-+0000` and `ros-hydro-image-geometry_1.10.18-0precise-20140613-1913-+0000_amd64` from ros-shadow-fixed:
  
  ```
  $ cmake --find-package -DNAME=OpenCV -DCOMPILER_ID=GNU -DLANGUAGE=CXX -DMODE=LINK
  -rdynamic /opt/ros/hydro/lib/libopencv_videostab.so.2.4.9 /opt/ros/hydro/lib/libopencv_video.so.2.4.9 /opt/ros/hydro/lib/libopencv_ts.a /opt/ros/hydro/lib/libopencv_superres.so.2.4.9 /opt/ros/hydro/lib/libopencv_stitching.so.2.4.9 /opt/ros/hydro/lib/libopencv_photo.so.2.4.9 /opt/ros/hydro/lib/libopencv_ocl.so.2.4.9 /opt/ros/hydro/lib/libopencv_objdetect.so.2.4.9 /opt/ros/hydro/lib/libopencv_nonfree.so.2.4.9 /opt/ros/hydro/lib/libopencv_ml.so.2.4.9 /opt/ros/hydro/lib/libopencv_legacy.so.2.4.9 /opt/ros/hydro/lib/libopencv_imgproc.so.2.4.9 /opt/ros/hydro/lib/libopencv_highgui.so.2.4.9 /opt/ros/hydro/lib/libopencv_gpu.so.2.4.9 /opt/ros/hydro/lib/libopencv_flann.so.2.4.9 /opt/ros/hydro/lib/libopencv_features2d.so.2.4.9 /opt/ros/hydro/lib/libopencv_core.so.2.4.9 /opt/ros/hydro/lib/libopencv_contrib.so.2.4.9 /opt/ros/hydro/lib/libopencv_calib3d.so.2.4.9 -ldl -lm -lpthread -lrt /opt/ros/hydro/lib/libopencv_nonfree.so.2.4.9 /opt/ros/hydro/lib/libopencv_ocl.so.2.4.9 /opt/ros/hydro/lib/libopencv_gpu.so.2.4.9 /opt/ros/hydro/lib/libopencv_photo.so.2.4.9 /opt/ros/hydro/lib/libopencv_objdetect.so.2.4.9 /opt/ros/hydro/lib/libopencv_legacy.so.2.4.9 /opt/ros/hydro/lib/libopencv_video.so.2.4.9 /opt/ros/hydro/lib/libopencv_ml.so.2.4.9 /opt/ros/hydro/lib/libopencv_calib3d.so.2.4.9 /opt/ros/hydro/lib/libopencv_features2d.so.2.4.9 /opt/ros/hydro/lib/libopencv_highgui.so.2.4.9 /opt/ros/hydro/lib/libopencv_imgproc.so.2.4.9 /opt/ros/hydro/lib/libopencv_flann.so.2.4.9 /opt/ros/hydro/lib/libopencv_core.so.2.4.9 -Wl,-rpath,/opt/ros/hydro/lib 
  ```
  
  ```
  $ cat /opt/ros/hydro/lib/pkgconfig/image_geometry.pc
  prefix=/opt/ros/hydro
  
  Name: image_geometry
  Description: Description of image_geometry
  Version: 1.10.18
  Cflags: -I/opt/ros/hydro/include -I/opt/ros/hydro/include/opencv
  Libs: -L/opt/ros/hydro/lib -limage_geometry
  Requires: sensor_msgs
  ```

No change in image_geometry between 1.10.17 and 1.10.18 could explain the missing opencv export, so it must be the versioned library names in the exported `OpenCV_LIBRARIES` cmake variable that `catkin_package(DEPENDS OpenCV2 ...)` in image_geometry's `CMakeLists.txt` cannot handle.

**Comment by vrabaud** [comment]
It is indeed in there and that is a bug in catkin that got fixed some time ago in https://github.com/ros/catkin/pull/616 but it has not been released since then. It will be on 16/06/2014, hence today :)
I'll wait to see if that truely fixes it and I'll handfix OpenCV otherwise.

**Comment by vrabaud** [comment]
hmm, no, you pointed out something different, will fix.

**Comment by vrabaud** [comment]
actually no, it is that one. Will wait for one day.

**Comment by vrabaud** [comment]
Those builds now work. Groovy will soon have the same fix. Sorry for the hassle.

**Comment by meyerj** [comment]
The groovy build jobs of [hector_worldmodel](http://wiki.ros.org/hector_worldmodel) are still failing due to OpenCV linker errors. I am not sure if they are caused by the same catkin bug than in hydro as hector_worldmodel is a dry stack in groovy, but at least the patch in ros/catkin#616 has not been merged to groovy-devel yet.
- http://jenkins.ros.org/view/GbinP32/job/ros-groovy-hector-worldmodel_binarydeb_precise_i386/
- http://jenkins.ros.org/view/GbinP64/job/ros-groovy-hector-worldmodel_binarydeb_precise_amd64/
- http://jenkins.ros.org/view/GbinQ32/job/ros-groovy-hector-worldmodel_binarydeb_quantal_i386/
- http://jenkins.ros.org/view/GbinQ64/job/ros-groovy-hector-worldmodel_binarydeb_quantal_amd64/
- http://jenkins.ros.org/view/GbinO32/job/ros-groovy-hector-worldmodel_binarydeb_oneiric_i386/
- http://jenkins.ros.org/view/GbinO64/job/ros-groovy-hector-worldmodel_binarydeb_oneiric_amd64/

**Comment by vrabaud** [comment]
They are due to the same bug: they will push the new catkin as soon as they fix a bug in the farm. It should definitely not have taken that long, sorry for the hassle, should be any day now.

**Comment by meyerj** [comment]
The problem still persists and I get almost daily email reminders due to failed builds of [hector_worldmodel](http://www.ros.org/debbuild/groovy.html?q=hector_worldmodel) from Jenkins. It would be nice to have version 0.2.3 binaries in the groovy repositories before the groovy build farm is shut down.

Any plans to release the catkin patch in groovy or should I update the hector_object_tracker package to depend on OpenCV directly?

**Comment by vrabaud** [comment]
@dirk-thomas  , any timeline on when the fixed catkin will be released for Groovy ? Thx.

**Comment by dirk-thomas** [comment]
Oh :worried: I barely get to the Hydro back ports nowadays. But I will plan some time early next week to actually do all pending Groovy backports. Sorry for the delay.

**Comment by dirk-thomas** [comment]
Finally done: ros/rosdistro#5328

**Comment by vrabaud** [comment]
yay ! thx @dirk-thomas  
@meyerj  , I do not see any problem anymore on http://www.ros.org/debbuild/groovy.html?q=regression , can you please confirm ? Thx.

**Comment by dirk-thomas** [comment]
I think @tfoote merged Groovy today with half a dozen regressions which haven't been fixed for long.

**Comment by meyerj** [comment]
@vrabaud I can confirm that hector_worldmodel builds without linker errors now with catkin 0.5.59. Many thanks!
